### ISSUE TITLE
Have Default Close Button

### ISSUE THREAD

**Issue by merlinpatt** [reactjs/react-modal#320]
### Summary

All the examples for react-modal have a close button that has to be added whenever you create a modal with this package.

It would be great to see a close button built into this package, which could also be overridden.

### Expected behavior

You could pass in arguments like so

```
const CloseButton = () => (<button>Close modal</button>);

class SimpleModal extends React.Component {
  ...
  render () {
    return (
      <div>
        <button onClick={this.handleOpenModal}>Trigger Modal</button>
        <ReactModal 
           isOpen={this.state.showModal}
           contentLabel="Minimal Modal Example"
           closeButton={CloseButton}
        >
        </ReactModal>
      </div>
    );
  }
}
```

Internally this could be handled with

```
    const CloseButton = this.props.closeButton;
    return this.shouldBeClosed() ? <div /> : (
      <div ...>
        <div ....>
          {this.props.children}
          <span onClick={this.handleCloseButtonMouseDown>
            <CloseButton />
          </span>
        </div>
      </div>
    );
```

**Comment by GarrettHartley** [comment]
I understand there is the option to set the modal to close if someone has clicked on the backdrop area. But is there no default close button implemented yet?

**Comment by merlinpatt** [comment]
The backdrop close is a nice option, but I've found that most users will click an X or a button that says close instead of clicking the backdrop. The backdrop close feels like it is more for techy people than for general users

**Comment by GarrettHartley** [comment]
I agree. Any plans to build a close button or X into the package?

**Comment by merlinpatt** [comment]
I'd be willing to write the code and make the PR for this. I would just like to get an OK before I do

**Comment by diasbruno** [comment]
@merlinpatt @GarrettHartley It would be great (and maybe it can be disabled passing `closeButton={null}` or use a custom button).

**Comment by diasbruno** [comment]
ping @claydiffrient...to see if we can add this to `v2.0` and thoughts about this feature.

**Comment by merlinpatt** [comment]
@diasbruno do you have an idea of `v2.0` might come out?

I'd like to see the option for this sooner, and I think that could be done by having there be no close button as default, but allow it to be enabled with a boolean `useDefaultCloseButton`.

And yes, my thoughts exactly on the ability to disable it or replace it with a custom button

**Comment by sgilligan** [comment]
In my usage I apply a close button as a UI element/behaviour of the end-component. Makes sense for an alert dialog to have a close, but a login modal might not (this is the behaviour of my login). So yes, disabling is something I'd need to be able to do.

**Comment by diasbruno** [comment]
@merlinpatt there is a milestone [v2.0](https://github.com/reactjs/react-modal/milestone/1), but that comment was just for 'scheduling' purpose, if this feature can get early into master, that would be great.

**Comment by claydiffrient** [comment]
I've got mixed feelings on this.

On one hand, I can see the benefit of having something like this as an out of the box awesome thing.  This would encourage people to have a close button, and be nice. On the other hand, I do like having the library be as general use as possible.  I think that @sgilligan's comment about a login box is definitely applicable.

I've seen the modal used as a popover component before and many other instances where perhaps a close button doesn't make sense.  I also am hesitant to put a generic close button in place for accessibility and internationalization purposes.  The most unobtrusive close button I can think of is a simple "X", but that's not great for a11y because you'd want a screenreader to read "Close".  The the trouble comes that we'd need to provide a way to have the label prop get applied because perhaps "Close" isn't what it needs to read, but instead "Fermer" (the French for "close"). So we then have a i18n problem because translations are something we won't provide out of the box. 

So then we provide a sub-par experience for either a11y or i18n out of the box which isn't our goal at all.  Sure people can replace the default button with a nicer button, but then why have a default at all if everyone is going to replace it with a better close button anyways?

**Comment by merlinpatt** [comment]
@claydiffrient Regarding a11y, I think that can be handled with either an aria-label like Foundation uses (see code below), or just have the X be stylized with the word close next to or below it (like Slack has with the word `esc`).

For i18n, I think that could be handled with a property like `closeAriaLabel` or `closeContentLabel`, similar to how you already have it for content.

In regards to your last question about why bother, I think that there's a good number of people who use  packages like this one and just do the defaults.

From http://foundation.zurb.com/sites/docs/reveal.html
```
<button class="close-button" data-close aria-label="Close modal" type="button">
  <span aria-hidden="true">&times;</span>
</button>
```

**Comment by claydiffrient** [comment]
Thanks for the input, @merlinpatt. Coming back to look at this again, react-modal is really a general use modal component that can be used in whatever way the consumer would like to use it in.  It does however have some emphasis on a11y so we do a few things that make it so that is the number one priority.  react-modal isn't really a dialog box or anything like that out of the box.  It leaves creative freedom up to the consumers of it.

I don't feel comfortable putting any sort of default close button as part of the modal.  I feel it would be overridden by almost every consumer of it.  It's easy enough to extend into another more specific component that includes a close button.  I think that is where this really belongs.  We aren't trying to specify the UI for our end users like a library like Bootstrap or Foundation does.

From the link you posted from Foundation:
> You'll also need a way to close the modal from inside. By default, modals will close if clicked outside of, or if the esc key is pressed. However, you'll generally also want to add your own click trigger. Add the attribute data-close to any element within the modal to add one.

So they don't add a default button to the modal, that's something that you have to do.  That's the same as what react-modal does.  Even with Bootstrap you have to add in your own close button in the code.  It's not something that defaults to being there just by having a single modal element.

In short, I don't think a default button belongs at this level.  We have demos showing a close button being used which is essentially what every other modal library I've seen does.

**Comment by diasbruno** [comment]
@merlinpatt @claydiffrient @sgilligan @GarrettHartley what if `react-modal` provides a `BasicModal` which support common functionalities? It could have:

- Close button on the top (or positioned with css)
- Header section (optional)
- Content section
- Fotter section (optional)

**Comment by diasbruno** [comment]
Reopened just for visibility.

**Comment by claydiffrient** [comment]
@diasbruno I'd rather have that be a separate project that consumes `react-modal` instead. 

This allows us to keep `react-modal` as a good base layer.  It also reduces the number of things that we have to maintain within the library, allowing us to focus on the core.

**Comment by diasbruno** [comment]
@claydiffrient yeah, good point.

**Comment by wanderingcoder2** [comment]
Can this be re-opened, when used with react form  this button causes issues because you did NOT put type='button' and thus it defaults to being a submit, can you please edit the close button to be type button since it is NEVER intended to be a submit.

**Comment by seanmadsen** [comment]
I like the idea of a separate project that consumes `react-modal`. Does such a project exist yet? If so, it would be great to drop a link here for otherr who happen to stumble upon this thread looking for a higher-level, slightly-more-polished modal component, like me.
